### PR TITLE
fix: sceneCoordsToViewportCoords, jumping text when there is an offset (#4413)

### LINKED ISSUE #4413: getViewportCoords and viewportCoordsToSceneCoords don't transform into each other
shouldn't this resolve to true?

```typescript
const [clientX,clientY] = getViewportCoords(x1,y1);
const { x2, y2 } = viewportCoordsToSceneCoords({clientX,clientY});
return x2===x1 && y2===y1;
```

and

```typescript
const { clientX, clientY } = sceneCoordsToViewportCoords (x1,y1);
const { x2, y2 } = viewportCoordsToSceneCoords({clientX,clientY});
return x2===x1 && y2===y1;
```

https://github.com/excalidraw/excalidraw/blob/98b5c37e452d56b7d0fdbce82ce31e6389ba5330/src/utils.ts#L209

https://github.com/excalidraw/excalidraw/blob/98b5c37e452d56b7d0fdbce82ce31e6389ba5330/src/utils.ts#L231

https://github.com/excalidraw/excalidraw/blob/98b5c37e452d56b7d0fdbce82ce31e6389ba5330/src/components/App.tsx#L1911

### PR DESCRIPTION
**Pull request by zsviczian**
This resolves #4413 and fixes the incorrect sceneCoordsToViewportCoords function.

# Before

https://user-images.githubusercontent.com/14358394/150638060-60ea5927-c48e-4977-8224-ea6c295ada30.mp4

# After

https://user-images.githubusercontent.com/14358394/150638067-68cecb61-dbae-469a-9384-d600351e41f4.mp4

### CODE REVIEW

**Comment by zsviczian** [comment]
@ad1992, this simplifies and resolves the errors (and heuristic "magic") with text positioning in textwysiwyg. This change greatly improves the user experience for Excalidraw when it is rendered embedded in an app with offsetLeft & offsetTop !=0.

Please take a look and let's push this important improvement to production.

File: src/element/textWysiwyg.tsx [hunk 1]
```diff
@@ -173,29 +171,12 @@ export const textWysiwyg = ({
...
       if (!container) {
-        maxWidth =
-          (appState.offsetLeft + appState.width - viewportX - 8) /
-            appState.zoom.value -
-          // margin-right of parent if any
-          Number(
-            getComputedStyle(
-              excalidrawContainer?.parentNode as Element,
-            ).marginRight.slice(0, -2),
-          );
```
Comment (dwelle) [inline src/element/textWysiwyg.tsx:184]: note sure about what this solved before, but shouldn't w keep this?

Comment (zsviczian) [reply src/element/textWysiwyg.tsx:184]: I removed this deliberately and after extensive testing. I think it is based on the misunderstanding that appState.width is the entire screen width, while in reality, it is the canvasDOMWidth. Subtracting the right-hand side margin is unnecessary. 

https://github.com/excalidraw/excalidraw/blob/c009e03c8e654f89739f840c5b374bbb9d1d2982/src/components/App.tsx#L380-L395

**Comment by ad1992** [comment]
This is a great improvement @zsviczian ❤️.  Could you also write specs for this ? Will do one round of testing and then should be good to merge.

**Comment by zsviczian** [comment]
> This is a great improvement @zsviczian ❤️. Could you also write specs for this ? Will do one round of testing and then should be good to merge.

Great! A second (and 3rd, 4th...) round of testing is always essential. What do you mean by specs?

**Comment by ad1992** [comment]
> > This is a great improvement @zsviczian ❤️. Could you also write specs for this ? Will do one round of testing and then should be good to merge.
> 
> Great! A second (and 3rd, 4th...) round of testing is always essential. What do you mean by specs?

I meant tests for the change just to make sure we don't break in future as this part has been broken often in the past.

**Comment by zsviczian** [comment]
> > > This is a great improvement @zsviczian ❤️. Could you also write specs for this ? Will do one round of testing and then should be good to merge.
> > 
> > 
> > Great! A second (and 3rd, 4th...) round of testing is always essential. What do you mean by specs?
> 
> I meant tests for the change just to make sure we don't break in future as this part has been broken often in the past.

I think this broke frequently in the future because there was a formula error. sceneCoordsToViewportCoords was wrong, and we were trying to resolve its impact by applying complex calculations in textwysiwyg.  That should be resolved for good.

Regarding testing... I don't really know how to test for zoom and offset situations... I haven't yet looked at excalidraw tests, I'll take a look to see if this sort of testing can be automated.  My recommendation would be to deploy a second app with every change, the second should have offsets and righthand side margin.

**Comment by ad1992** [comment]
> > > > This is a great improvement @zsviczian ❤️. Could you also write specs for this ? Will do one round of testing and then should be good to merge.
> > > 
> > > 
> > > Great! A second (and 3rd, 4th...) round of testing is always essential. What do you mean by specs?
> > 
> > 
> > I meant tests for the change just to make sure we don't break in future as this part has been broken often in the past.
> 
> I think this broke frequently in the future because there was a formula error. sceneCoordsToViewportCoords was wrong, and we were trying to resolve its impact by applying complex calculations in textwysiwyg. That should be resolved for good.
> 
> Regarding testing... I don't really know how to test for zoom and offset situations... I haven't yet looked at excalidraw tests, I'll take a look to see if this sort of testing can be automated.

Sure 👍🏻 .  The tests would probably be tricky for this but lets give it a try and decide accordingly.

>My recommendation would be to deploy a second app with every change, the second should have offsets and righthand side margin.

What do you mean by deploying an app with offsets and right handed margin ? We have the same code sandbox example running with package example in port 3001 which can be used for testing

**Comment by zsviczian** [comment]
> What do you mean by deploying an app with offsets and right handed margin ? We have the same code sandbox example running with package example in port 3001 which can be used for testing

To make the question concrete. If I would want to test a PR with screen offsets, how can I do that. For sake of example, the current preview release for the hyperlinks PR is https://excalidraw-f13roaq8h-excalidraw.vercel.app/ I tried https://excalidraw-f13roaq8h-excalidraw.vercel.app:3001/ but that did not bring up anything...

**Comment by ad1992** [comment]
> > What do you mean by deploying an app with offsets and right handed margin ? We have the same code sandbox example running with package example in port 3001 which can be used for testing
> 
> To make the question concrete. If I would want to test a PR with screen offsets, how can I do that. For sake of example, the current preview release for the hyperlinks PR is https://excalidraw-f13roaq8h-excalidraw.vercel.app/ I tried https://excalidraw-f13roaq8h-excalidraw.vercel.app:3001/ but that did not bring up anything...

Ohh ok I meant it will work in [`localhost:3001`](https://github.com/excalidraw/excalidraw/blob/master/src/packages/excalidraw/README_NEXT.md#development), we can also deploy it as well so its available in preview too

**Comment by zsviczian** [comment]
> Ohh ok I meant it will work in [`localhost:3001`](https://github.com/excalidraw/excalidraw/blob/master/src/packages/excalidraw/README_NEXT.md#development), we can also deploy it as well so its available in preview too
That would be great! I was looking around the repository but didn't yet have the patience to figure it out. If you give me some pointers I am also happy to propose a PR with port 3001 being available in preview too.

----

BTW Have you had time to test this fix? I've deployed this to the Obsidian Excalidraw build a week ago and have not experienced any issues since. I am pretty confident that this is working as expected. Would be good to get this deployed before further functions build on the current erroneous sceneCoordsToViewportCoords() function.

**Review by ad1992 — APPROVED** [review]
+1 from my side, tested it out and it works great, thanks @zsviczian for fixing this 💯 , lets wait for @dwelle's +1 and then we are good to 🚀

**Comment by dwelle** [comment]
For sure, let's merge! Thanks!

**Review by dwelle — APPROVED** [review]

**Pull request merged**